File: ios/Classes/Data/InAppMessageCoder.swift
//
//  InAppMessageCoder.swift
//  exponea
//

import Foundation
import ExponeaSDK

class InAppMessageCoder {
    static func encode(_ source: InAppMessage) throws -> [String:Any?] {
        return [
            "id": source.id,
            "name": source.name,
            "messageType": source.messageType.rawValue,
            "frequency": source.frequency?.rawValue,
            "payload": source.payload != nil ? try? String(data: JSONEncoder().encode(source.payload), encoding: .utf8) : nil,
            "variantId": source.variantId,
            "variantName": source.variantName,
            "trigger": try? String(data: JSONEncoder().encode(source.trigger), encoding: .utf8),
            "dateFilter": try? String(data: JSONEncoder().encode(source.dateFilter), encoding: .utf8),
            "loadPriority": source.priority,
            "loadDelay": source.delayMS,
            "closeTimeout": source.timeoutMS,
            "payloadHtml": source.payloadHtml,
            "isHtml": source.isHtml,
            "hasTrackingConsent": source.hasTrackingConsent,
            "consentCategoryTracking": source.consentCategoryTracking,
        ]
    }
    
    static func decode(_ source: [String:Any?]) throws -> InAppMessage {
        var payload: InAppMessagePayload? = nil
        if let payloadString: String = try? source.getOptional("payload"),
           let payloadData: Data = payloadString.data(using: .utf8)
        {
            payload = try? JSONDecoder().decode(InAppMessagePayload.self, from: payloadData)
        }
        
        guard let triggerString: String = try source.getRequired("trigger"),
              let triggerData: Data = triggerString.data(using: .utf8),
              let trigger = try? JSONDecoder().decode(EventFilter.self, from: triggerData) else {
            throw ExponeaDataError.missingProperty(property: "trigger")
        }
        
        guard let dateFilterString: String = try source.getRequired("dateFilter"),
              let dateFilterData: Data = dateFilterString.data(using: .utf8),
              let dateFilter = try? JSONDecoder().decode(DateFilter.self, from: dateFilterData) else {
            throw ExponeaDataError.missingProperty(property: "dateFilter")
        }
        
        return InAppMessage(
            id: try source.getRequired("id"),
            name: try source.getRequired("name"),
            rawMessageType: try source.getOptional("messageType"),
            rawFrequency: try source.getRequired("frequency"),
            payload: payload,
            variantId: try source.getRequired("variantId"),
            variantName: try source.getRequired("variantName"),
            trigger: trigger,
            dateFilter: dateFilter,
            priority: try source.getOptional("loadPriority"),
            delayMS: try source.getOptional("loadDelay"),
            timeoutMS: try source.getOptional("closeTimeout"),
            payloadHtml: try source.getOptional("payloadHtml"),
            isHtml: try source.getOptional("isHtml"),
            hasTrackingConsent: try source.getOptional("hasTrackingConsent"),
            consentCategoryTracking: try source.getOptional("consentCategoryTracking")
        )
    }
}

File: android/src/main/kotlin/com/exponea/data/InAppMessageCoder.kt
package com.exponea.data

import com.exponea.getRequired
import com.exponea.sdk.models.DateFilter
import com.exponea.sdk.models.InAppMessage
import com.exponea.sdk.models.InAppMessagePayload
import com.exponea.sdk.models.eventfilter.EventFilter
import com.exponea.sdk.util.ExponeaGson

class InAppMessageCoder {
    companion object {
        fun encode(source: InAppMessage): Map<String, Any?> {
            return mapOf(
                "id" to source.id,
                "name" to source.name,
                "messageType" to source.messageType.value,
                "frequency" to source.frequency?.value,
                "payload" to source.payload?.let {
                    ExponeaGson.instance.toJson(it)
                },
                "variantId" to source.variantId,
                "variantName" to source.variantName,
                "trigger" to source.trigger?.let {
                    ExponeaGson.instance.toJson(it)
                },
                "dateFilter" to ExponeaGson.instance.toJson(source.dateFilter),
                "loadPriority" to source.priority,
                "loadDelay" to source.delay,
                "closeTimeout" to source.timeout,
                "payloadHtml" to source.payloadHtml,
                "isHtml" to source.isHtml,
                "hasTrackingConsent" to source.hasTrackingConsent,
                "consentCategoryTracking" to source.consentCategoryTracking,
                "isRichText" to source.isRichText,
            )
        }

        fun decode(source: Map<String, Any?>): InAppMessage {
            return InAppMessage(
                id = source.getRequired("id"),
                name = source.getRequired("name"),
                rawMessageType = source.getOptional("messageType"),
                rawFrequency = source.getRequired("frequency"),
                payload = ExponeaGson.instance.fromJson(source.getOptional<String>("payload"), InAppMessagePayload::class.java),
                variantId = source.getRequired("variantId"),
                variantName = source.getRequired("variantName"),
                trigger = ExponeaGson.instance.fromJson(source.getOptional<String>("trigger"), EventFilter::class.java),
                dateFilter = ExponeaGson.instance.fromJson(source.getOptional<String>("dateFilter"), DateFilter::class.java),
                priority = source.getOptional("loadPriority"),
                delay = source.getOptional<Int>("loadDelay")?.toLong(),
                timeout = source.getOptional<Int>("closeTimeout")?.toLong(),
                payloadHtml = source.getOptional("payloadHtml"),
                isHtml = source.getOptional("isHtml"),
                rawHasTrackingConsent = source.getOptional("hasTrackingConsent"),
                consentCategoryTracking = source.getOptional("consentCategoryTracking"),
                isRichText = source.getOptional("isRichText"),
            )
        }
    }
}